syntax = "proto3";

package users;

service Users {
    rpc GetUsers(GetUsersRequest) returns (GetUsersResponse) {}
    rpc GetUserById(GetUserByIdRequest) returns (GetUserByIdResponse) {}
    rpc CreateUser(CreateUserRequest) returns (CreateUserResponse) {}
    rpc UpdateUser(UpdateUserRequest) returns (UpdateUserResponse) {}
    rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse) {}
}

message User {
    int32 id = 1;
    string name = 2;
    string email = 3;
    string password = 4;
}

message GetUsersRequest {
}

message GetUsersResponse {
    repeated User users = 1;
}

message GetUserByIdRequest {
    int32 id = 1;
}

message GetUserByIdResponse {
    User user = 1;
}

message CreateUserRequest {
    string name = 1;
    string email = 2;
}

message CreateUserResponse {
    User user = 1;
}

message UpdateUserRequest {
    int32 id = 1;
    optional string name = 2;
    optional string email = 3;
    optional string password = 4;
}

message UpdateUserResponse {
    User user = 1;
}

message DeleteUserRequest {
    int32 id = 1;
}

message DeleteUserResponse {
    bool success = 1;
}
